### PR TITLE
[SPARK-21342] Fix DownloadCallback to work well with RetryingBlockFetcher.

### PR DESCRIPTION
**Pull request by jinxing64**
## What changes were proposed in this pull request?

When `RetryingBlockFetcher` retries fetching blocks. There could be two `DownloadCallback`s download the same content to the same target file. It could cause `ShuffleBlockFetcherIterator` reading a partial result.

This pr proposes to create and delete the tmp files in `OneForOneBlockFetcher`

### CODE REVIEW

**Comment by jinxing64** [comment]
cc @zsxwing @cloud-fan @jiangxb1987

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java [hunk 1]
```diff
@@ -151,15 +152,27 @@ private void failRemainingBlocks(String[] failedBlockIds, Throwable e) {
...
 
     private WritableByteChannel channel = null;
     private File targetFile = null;
+    private File tmpFile = null;
     private int chunkIndex;
 
-    DownloadCallback(File targetFile, int chunkIndex) throws IOException {
-      this.targetFile = targetFile;
-      this.channel = Channels.newChannel(new FileOutputStream(targetFile));
+    DownloadCallback(int chunkIndex) throws IOException {
```
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: This is unrelated, just remove a redundant param.

**Review by zsxwing — CHANGES_REQUESTED** [review]
Made one pass.

I also noticed one potential file leak: if a file is downloaded after `ShuffleBlockFetcherIterator.cleanup` is called, they won't be deleted.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java [hunk 2]
```diff
@@ -151,15 +152,27 @@ private void failRemainingBlocks(String[] failedBlockIds, Throwable e) {
     }
   }
 
+  private static synchronized boolean renameFile(File src, File dest) {
+    if (dest.exists()) {
+      if (!dest.delete()) {
```
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: I was using `synchronized static` because there could be multiple `OneForOneBlockFetcher` doing the rename work.

**Comment by jinxing64** [comment]
There are some corner cases because of creating and deleting in both `ShuffleBlockFetcherIterator` and `OneForOneBlockFetcher`. It seems no need to pass the `shuffleFiles` from `ShuffleBlockFetcherIterator` to `ShuffleClient` and `OneForOneBlockFetcher`. Now I changed create and delete files only in `OneForOneBlockFetcher`. We don't do renaming. Every `DownloadCallback` will have their own target file. All target files will be deleted when blockManager stop.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java [hunk 3]
```diff
@@ -91,15 +95,17 @@ public void fetchBlocks(
       String execId,
       String[] blockIds,
       BlockFetchingListener listener,
-      File[] shuffleFiles) {
+      boolean toDisk) {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java]: shall we pass in `TmpFileCreater` here?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java [hunk 4]
```diff
@@ -91,15 +95,18 @@ public void fetchBlocks(
       String execId,
       String[] blockIds,
       BlockFetchingListener listener,
-      File[] shuffleFiles) {
+      boolean toDisk,
+      TmpFileCreater tmpFileCreater) {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java]: we can use null `tmpFileCreater` to represent not fetching to disk

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java [hunk 5]
```diff
@@ -91,15 +95,18 @@ public void fetchBlocks(
...
     logger.debug("External shuffle fetch from {}:{} (executor id {})", host, port, execId);
     try {
       RetryingBlockFetcher.BlockFetchStarter blockFetchStarter =
           (blockIds1, listener1) -> {
             TransportClient client = clientFactory.createClient(host, port);
-            new OneForOneBlockFetcher(client, appId, execId, blockIds1, listener1, conf,
-              shuffleFiles).start();
+            OneForOneBlockFetcher blockFetcher = new OneForOneBlockFetcher(client, appId, execId,
+              blockIds1, listener1, conf, toDisk, tmpFileCreater);
+            blockfetchers.add(blockFetcher);
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java]: `OneForOneBlockFetcher` doesn't implement `equals` and `hashCode`, so 2 `OneForOneBlockFetcher`s are considered equal if and only if they are same instance.

Here `blockfetchers.add(blockFetcher)` always add a new instance, so I don't thing using set to store block fetchers makes sense.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleClient.java [hunk 6]
```diff
@@ -42,5 +42,13 @@ public abstract void fetchBlocks(
...
       String[] blockIds,
       BlockFetchingListener listener,
-      File[] shuffleFiles);
+      boolean toDisk,
+      TmpFileCreater tmpFileCreater);
+
+  /**
+   * An interface to provide approach to create tmp file.
+   */
+  public interface TmpFileCreater {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleClient.java]: shall we move it to a separated file? We may use it outside of `ShuffleClient` in the future.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleClient.java [hunk 7]
```diff
@@ -42,5 +42,13 @@ public abstract void fetchBlocks(
...
+    File createTempBlock();
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleClient.java]: `createTempFile` may be a better name?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java [hunk 8]
```diff
@@ -151,14 +167,21 @@ private void failRemainingBlocks(String[] failedBlockIds, Throwable e) {
...
 
     private WritableByteChannel channel = null;
     private File targetFile = null;
     private int chunkIndex;
 
-    DownloadCallback(File targetFile, int chunkIndex) throws IOException {
-      this.targetFile = targetFile;
+    DownloadCallback(int chunkIndex) throws IOException {
+      this.targetFile = tmpFileCreater.createTempBlock();
+      shuffleFiles.add(targetFile);
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: ditto, we will create unique temp files, so `shuffleFiles` can be a list instead of set.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleClient.java (hunk 6 above)
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleClient.java]: Actually we may not need this, just let `ExternalShuffleClient.fetchBlocks` accept a java lambda.

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 9]
```diff
@@ -53,6 +53,7 @@ private[spark] class NettyBlockTransferService(
   private val serializer = new JavaSerializer(conf)
   private val authEnabled = securityManager.isAuthenticationEnabled()
   private val transportConf = SparkTransportConf.fromSparkConf(conf, "shuffle", numCores)
+  private val blockFetchers = collection.mutable.HashSet[OneForOneBlockFetcher]()
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala]: this is not thread safe. And it's a memory leak because it won't be cleaned until the executor is shutting down.

**Review by zsxwing — CHANGES_REQUESTED** [review]
> Every DownloadCallback will have their own target file.

This is much better.

> All target files will be deleted when blockManager stop.

The can be replaced with `Utils.createTempDir`. However, I'm not convinced that this is the right way. An executor could run for a long time and the size of these shuffle files is non-trivial.

An alternative way is using `ShuffleBlockFetcherIterator.isZombie` to check if you need to clean the shuffle file in the callback. You probably need to pass `ShuffleBlockFetcherIterator` to the callback as you may need to run some codes in the `ShuffleBlockFetcherIterator.synchronized` block.

**Comment by jinxing64** [comment]
In current change:
1. I'm using `java.util.function.Supplier` instead of `TmpFileCreater`
2. Pass `shuffleBlockFetcherIteratorIsZombie` from `ShuffleBlockFetcherIterator` to `OneForOneBlockFetcher`;
3. Succeeded files will be cleanup in `ShuffleBlockFetcherIterator`, while others will be deleted in `OneForOneBlockFetcher`

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java [hunk 10]
```diff
@@ -58,29 +59,41 @@
...
     this.chunkCallback = new ChunkCallback();
     this.transportConf = transportConf;
-    if (shuffleFiles != null) {
-      this.shuffleFiles = shuffleFiles;
-      assert this.shuffleFiles.length == blockIds.length:
-        "Number of shuffle files should equal to blocks";
-    }
+    this.toDisk = toDisk;
+    this.tmpFileCreater = tmpFileCreater;
+    this.shuffleBlockFetcherIteratorIsZombie = shuffleBlockFetcherIteratorIsZombie;
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: shall we add an assert that, if `toDisk` is true, `tmpFileCreater` and `shuffleBlockFetcherIteratorIsZombie` can not be null. Or alternatively, we can drop the `toDisk` parameter, and use `toDisk = tmpFileCreater != null`

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java [hunk 11]
```diff
@@ -58,29 +59,41 @@
...
       TransportClient client,
       String appId,
       String execId,
       String[] blockIds,
       BlockFetchingListener listener,
       TransportConf transportConf,
-      File[] shuffleFiles) {
+      Boolean toDisk,
+      Supplier<File> tmpFileCreater,
+      Supplier<Boolean> shuffleBlockFetcherIteratorIsZombie) {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: Since the network module doesn't depend on the core module, we should hide the `ShuffleBlockFetcherIterator` from network module. How about we name it `canCallerSideDeleteFile`?

**Comment by cloud-fan** [comment]
LGTM except 2 minor comments

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java [hunk 12]
```diff
@@ -119,9 +132,9 @@ public void onSuccess(ByteBuffer response) {
           // Immediately request all chunks -- we expect that the total size of the request is
           // reasonable due to higher level chunking in [[ShuffleBlockFetcherIterator]].
           for (int i = 0; i < streamHandle.numChunks; i++) {
-            if (shuffleFiles != null) {
+            if (toDisk) {
```
Comment (jiangxb1987) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: We can use `tmpFileCreater != null` instead, so we may remove all the `toDisk` params.

Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: Good idea.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java (hunk 11 above)
Comment (jinxing64) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: I will refine.

**Comment by zsxwing** [comment]
@jinxing64 I submitted https://github.com/jinxing64/spark/pull/1 to your repo to fix a potential file leak. Otherwise, this looks good to me.

**Comment by jinxing64** [comment]
Jenkins, retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master/2.2!
